b2sdk/sync/sync.py

```diff
@@ -180,52 +96,107 @@ def count_files(local_folder, reporter):
+        :type compare_version_mode: int
+        :param compare_threshold: should be greater than 0, default is 0
+        :type compare_threshold: int
+        :param keep_days: if keep_days_or_delete is 302, then this should be greater than 0
+        :type keep_days: int
+        """
+        self.newer_file_mode = newer_file_mode
+        self.keep_days_or_delete = keep_days_or_delete
+        self.keep_days = keep_days
+        self.compare_version_mode = compare_version_mode or AbstractFileSyncPolicy.COMPARE_VERSION_MODTIME
+        self.compare_threshold = compare_threshold or 0
+        self.dry_run = dry_run
+        self.allow_empty_source = allow_empty_source
+        self.policies_manager = policies_manager
+        self.max_workers = max_workers
+        self.no_progress = no_progress
+        self._validate()
+
+    def _validate(self):
+        if self.compare_threshold < 0:
+            raise InvalidArgument('compare_threshold', 'must be a positive integer')
+
+        if self.newer_file_mode and self.newer_file_mode not in AbstractFileSyncPolicy.NEWER_FILE_MODES:
+            raise InvalidArgument(
+                'newer_file_mode',
+                'must be one of :%s' % AbstractFileSyncPolicy.NEWER_FILE_MODES,
+            )
+
+        if self.keep_days_or_delete and self.keep_days_or_delete not in self.KEEP_OR_DELETE_MODES:
+            raise InvalidArgument(
+                'keep_days_or_delete',
+                'must be one of :%s' % self.KEEP_OR_DELETE_MODES,
+            )
+
+        if self.keep_days_or_delete == self.KEEP_DAYS_MODE and self.keep_days is None:
+            raise InvalidArgument(
+                'keep_days',
+                'is required when keep_days_or_delete is %s' % self.KEEP_DAYS_MODE,
+            )
+
+        if self.compare_version_mode and self.compare_version_mode not in AbstractFileSyncPolicy.COMPARE_VERSION_MODES:
+            raise InvalidArgument(
+                'compare_version_mode',
+                'must be one of :%s' % AbstractFileSyncPolicy.COMPARE_VERSION_MODES,
+            )
+
+    def sync_folders(self, source_folder, dest_folder, now_millis, reporter):
+        """
+        Syncs two folders.  Always ensures that every file in the
+        source is also in the destination.  Deletes any file versions
+        in the destination older than history_days.
+
+        :param source_folder: source folder object
+        :type source_folder: b2sdk.sync.folder.AbstractFolder
+        :param dest_folder: destination folder object
+        :type dest_folder: b2sdk.sync.folder.AbstractFolder
+        :param now_millis: current time in milliseconds
+        :type now_millis: int
+        """
+        # For downloads, make sure that the target directory is there.
+        if dest_folder.folder_type() == 'local' and not self.dry_run:
+            dest_folder.ensure_present()
+
+        if source_folder.folder_type() == 'local' and not self.allow_empty_source:
```

Comment:
I forget... Does `allow_empty_source` only apply to local sources, and not to syncing from B2 to local?

Reply:
this code was just moved from one place to another, haven't thought about this in a while. It looks like it may be a bug or intended behavior, no idea. I'm fine with it as it is - if you think it is wrong, then I suggest that we discuss it in a separate issue.

Reply:
I agree that it's a separate issue.  It's probably always been this way.  (And I probably wrote it.)